File: GolangSolutions/Array/M3164_优质数对的总数II.go
/*
	@author Hyperspace
	@date 2024/10/11
	@file M3164_优质数对的总数II.go
*/

package Array

func numberOfPairs(nums1 []int, nums2 []int, k int) (ans int64) {
	mp := map[int]int{}
	for _, v := range nums1 {
		if v%k != 0 {
			continue
		}
		v /= k
		for d := 1; d*d <= v; d++ {
			if v%d > 0 {
				continue
			}
			mp[d]++
			if d*d < v {
				mp[v/d]++
			}
		}
	}

	for _, v := range nums2 {
		ans += int64(mp[v])
	}

	return
}

File: CPPSolutions/LeetCode/Array/M3164_优质数对的总数II.cpp
#include <bits/stdc++.h>

using namespace std;

typedef long long LL;

/**
 * @author Hyper
 * @date 2024/10/11
 * @file M3164_优质数对的总数II.cpp
 */
class Solution {
public:
    LL numberOfPairs(vector<int> &nums1, vector<int> &nums2, int k) {
        unordered_map<int, int> mp;
        for (int v: nums1) {
            if (v % k) continue;
            v /= k;
            for (int d = 1; d * d <= v; d++) {
                if (v % d) continue;
                mp[d]++;
                if (d * d < v) mp[v / d]++;
            }
        }
        LL ans = 0;
        for (int v: nums2) ans += mp[v];

        return ans;
    }
};